File: bigseqkit/rmdup.go
package bigseqkit

import (
	"fmt"
	"ignis/driver/api"
	"ignis/executor/api/ipair"
)

type SeqKitRmDupOptions struct {
	inner RmDupOptions
}

type RmDupOptions struct {
	Config             KitConfig
	ByName             *bool
	BySeq              *bool
	IgnoreCase         *bool
	DupSeqsFile        *string
	DupNumFile         *string
	OnlyPositiveStrand *bool
}

func (this *RmDupOptions) setDefaults() *RmDupOptions {
	this.Config.setDefaults()
	setDefault(&this.ByName, false)
	setDefault(&this.BySeq, false)
	setDefault(&this.IgnoreCase, false)
	setDefault(&this.DupSeqsFile, "")
	setDefault(&this.DupNumFile, "")
	setDefault(&this.OnlyPositiveStrand, false)

	return this
}

func (this *SeqKitRmDupOptions) Config(v *SeqKitConfig) *SeqKitRmDupOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitRmDupOptions) ByName(v bool) *SeqKitRmDupOptions {
	this.inner.ByName = &v
	return this
}

func (this *SeqKitRmDupOptions) BySeq(v bool) *SeqKitRmDupOptions {
	this.inner.BySeq = &v
	return this
}

func (this *SeqKitRmDupOptions) IgnoreCase(v bool) *SeqKitRmDupOptions {
	this.inner.IgnoreCase = &v
	return this
}

func (this *SeqKitRmDupOptions) DupSeqsFile(v string) *SeqKitRmDupOptions {
	this.inner.DupSeqsFile = &v
	return this
}

func (this *SeqKitRmDupOptions) DupNumFile(v string) *SeqKitRmDupOptions {
	this.inner.DupNumFile = &v
	return this
}

func (this *SeqKitRmDupOptions) OnlyPositiveStrand(v bool) *SeqKitRmDupOptions {
	this.inner.OnlyPositiveStrand = &v
	return this
}

func RmDup(input *api.IDataFrame[string], o *SeqKitRmDupOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitRmDupOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	revcom := !*opts.OnlyPositiveStrand

	if *opts.BySeq && *opts.ByName {
		return nil, fmt.Errorf("only one/none of the flags -s (--by-seq) and -n (--by-name) is allowed")
	}

	if !revcom && !*opts.BySeq {
		return nil, fmt.Errorf("flag -s (--by-seq) needed when using -P (--only-positive-strand)")
	}

	prepare, err := api.AddParam(libSource("RmDupPrepare"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	prepared, err := api.MapPartitions[string, ipair.IPair[int64, string]](input, prepare)
	if err != nil {
		return nil, err
	}

	grouped, err := api.GroupByKey[int64, string](api.ToPair[int64, string](prepared), nil)
	if err != nil {
		return nil, err
	}

	check, err := api.AddParam(libSource("RmDupCheck"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	return api.Flatmap[ipair.IPair[int64, []string], string](grouped.FromPair(), check)
}

File: bigseqkit-py/bigseqkit/rmdup.py
from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitRmDupOptions:

    def __init__(self):
        self.__inner = RmDupOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def byName(self, v: bool):
        self.__inner.ByName = v

    def bySeq(self, v: bool):
        self.__inner.BySeq = v

    def ignoreCase(self, v: bool):
        self.__inner.IgnoreCase = v

    def dupSeqsFile(self, v: str):
        self.__inner.DupSeqsFile = v

    def dupNumFile(self, v: str):
        self.__inner.DupNumFile = v

    def onlyPositiveStrand(self, v: bool):
        self.__inner.OnlyPositiveStrand = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        revcom = not opts.OnlyPositiveStrand

        if opts.BySeq and opts.ByName:
            raise RuntimeError("only one/none of the flags -s (--by-seq) and -n (--by-name) is allowed")

        if not revcom and not opts.BySeq:
            raise RuntimeError("flag -s (--by-seq) needed when using -P (--only-positive-strand)")

        prepare = _libSource("RmDupPrepare").addParam("opts", _optionsToString(opts))

        prepared = input.mapPartitions(prepare)
        grouped = prepared.toPair().groupByKey()

        check = _libSource("RmDupCheck").addParam("opts", _optionsToString(opts))
        return grouped.flatmap(check)

class RmDupOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.ByName = None  # bool
        self.BySeq = None  # bool
        self.IgnoreCase = None  # bool
        self.DupSeqsFile = None  # string
        self.DupNumFile = None  # string
        self.OnlyPositiveStrand = None  # bool

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "ByName", False)
        _setDefault(self, "BySeq", False)
        _setDefault(self, "IgnoreCase", False)
        _setDefault(self, "DupSeqsFile", "")
        _setDefault(self, "DupNumFile", "")
        _setDefault(self, "OnlyPositiveStrand", False)


def rmDup(input: IDataFrame, o: SeqKitRmDupOptions = None, **kwargs):
    if o is None:
        o = SeqKitRmDupOptions()
    return o._run(input, **kwargs)
